### DISCUSSION TITLE
New Prometheus query editor (monaco) no longer support CTRL + ENTER

### DISCUSSION THREAD

**Discussion by cykl** [grafana/grafana#44010] [Observability]
**What happened**:

Most Grafana query editors  allow to re-run the query by hitting CTRL + ENTER. 

Prometheus query editor supported this shortcut until Grafana 8.2. Since Grafana 8.3, it no longer does :sob: 

Tweaking PromQL queries used to be a rather smooth experience. It is now a pain is the ass as you have to leave the keyboard, move the mouse to the top right corner to click on re-run, move the mouse back to give focus to the editor and grab the keyboard again and again. It's so infuriating that I ended up leaving Grafana explore view yesterday to use Prometheus UI instead :disappointed: 

**What you expected to happen**:

Bring back CTRL + ENTER! (or any other shortcut that work consistently across data source) 

**How to reproduce it (as minimally and precisely as possible)**:

1. Open Explore view
2. Select a Prometheus data source
3. Edit a query
4. Hit CTRL + ENTER. Observe nothing happen and you must use the mouse to re-run the query.

**Anything else we need to know?**:

No. 

**Environment**:
- Grafana version: 8.3.3
- Data source type & version: Prometheus
- User OS & Browser: Firefox 94

**Comment by zuchka** [comment]
@cykl 

can you please try and replicate this on our public sandbox instance, https://play.grafana.org?

For me, with play running Grafana v8.3.3 (30bb7a93c), CTRL + ENTER ~~still works~~ does not work in Explore and in the panel query editor. 

I tested on Firefox 93.0 and Chrome 96.0.4664.93 (Official Build) (arm64)

update: SHIFT + ENTER still works

**Comment by cykl** [comment]
@zuchka Good to know that CTRL + ENTER should still be there :relieved: 

play.grafana.com behaves the same than my local instance.

https://user-images.githubusercontent.com/1764725/147199937-8b620095-1d42-4dfd-920a-087bf4f164e7.mp4

(I tested in private mode to disable most plugins just to be safe)

Since this seems browser related, my detailed Browser info are:

 - Name 	Firefox
 - Version 	95.0
 - Build ID 	20211206150432
 - Distribution ID 	fedora
 - User Agent 	Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:95.0) Gecko/20100101 Firefox/95.0
 - OS Linux 5.15.10-200.fc35.x86_64 #1 SMP Fri Dec 17 14:46:39 UTC 2021

I think it's unrelated, but please note than I use an US international keyboard layout (i.e. with dead keys) which caused problems in the previous editor. 

I will try to gather more data points from coworkers, maybe we can find a common pattern.

**Comment by cykl** [comment]
One co-worker might have figured it out. Loki  and old Prometheus query editor supported both shift + enter and ctrl + enter. New Prometheus only supports shift + enter. 

I'm used to ctrl + enter as it matches the behavior of many tools (ex. enter adds a new line, ctrl + enter sends a message). 

I tried to find a list of all keyboard shortcuts in the documentation before raising this issue and only found https://grafana.com/docs/grafana/latest/dashboards/shortcuts/. Might be a good addition to the docs.

**Comment by zuchka** [comment]
@cykl apologies. My muscle memory is so ingrained using explorer that I actually tested this with SHIFT and not CTRL. CTRL + ENTER didn't work for me on play either. Apologies for that confusion, but I'm glad the SHIFT shortcut still works.

And YES, I don't even know all the hotkeys and shortcuts in Grafana. @grafana/docs-squad, should we add all the available shortcuts to the docs? Right now you can only view them by pressing `h` inside a grafana instance:

![CleanShot 2021-12-23 at 08 56 42@2x](https://user-images.githubusercontent.com/37156449/147281902-fd5401c6-7176-4c35-8ed8-3ade8e1097a1.png)

The additional confusion is that the SHIFT + ENTER for running queries appears specific to several plugins like the Prometheus datasource...so maybe a note in the Prometheus data source docs? thoughts?

**Comment by aocenas** [comment]
At this point, I think it makes sense to move this to the discussions as it's more about whether it makes more sense to have ctrl+enter or shift+enter and we may also have some inconsistencies as each data source needs to make sure it handles this correctly.

**Comment by SethFalco** [comment]
Sorry to comment on an old discussion, just wanted to chime in as this was bothering me a bit!

In many apps doing `CTRL` + `ENTER` acts as a "finished" action to either send, post, or save whatever the user is typing.

For example:
* To send an email in email clients like Thunderbird and Tutanota.
* Submitting issues, pull requests, and comments on GitHub and GitLab.
* Commenting on Reddit.
* Tweeting on Twitter.

Currently, the user can do `SHIFT` + `ENTER` to submit a query, but this isn't very intuitive. The only reason I found out about that shortcut is because I was looking in that area of the repository. ^-^'

Would anyone have anything against also making `CTRL` + `ENTER` an acceptable way to submit the query?
The current behavior with `SHIFT` + `ENTER` should probably be preserved for Grafana users that are already accustomed to using that shortcut, however.

**Comment by krisztianfekete** [comment]
One thing that is driving me crazy is that Prometheus' UI is executing the query after I hit Enter, while Grafana's editor is starting a new line. I am using both applications frequently, and I would prefer to not have this inconsistency. IIRC years ago the applications were behaving the same way regarding this feature.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #83022. Feel free to continue the discussion around this there. Thank you!
